### PR TITLE
Add woocommerce_output_product_categories so we can handle legacy templates.

### LINKED ISSUE #18674: [3.3] Categories showing on shop page when the shouldn't be.
Creating this as a parent issue for tracking this common reported issue. There seems to be different causes, sometimes plugins (Genesis Connect) and sometimes themes (Flatsome). Still trying to find the exact change that caused this.

While technically this the 3rd party conflict territory, it's something worth looking into to see if there is a bandaid that could be applied.

Some examples: 

- https://wordpress.org/support/topic/urgent-issues-because-of-todays-update/
- https://wordpress.org/support/topic/problem-shop-page/
- https://github.com/woocommerce/woocommerce/issues/18673

_____

If this is happening for you, please don't open a new issue or reply with just a "happening for me too". Instead, please try to find which plugin or theme is causing this to happen on your site, and let us know.

Also, double check and make sure the settings are correctly configured in the customizer: http://cld.wthms.co/1Xi59n

### PR DESCRIPTION
**Pull request by mikejolley**
This is a tricky one to workaround.

In 3.2 the product subcategory function checks the global loop and outputs categories. This is called from template functions. https://github.com/woocommerce/woocommerce/blob/3.2.6/includes/wc-template-functions.php#L1750-L1782

3.3 replaces the template function calls with `woocommerce_maybe_show_product_subcategories` which is all hooked in. Templates need to do nothing.

The issue is, if an old template is being used, it will just render cats with the default parent 0. This is what is causing issues.

Themes updating template files with resolve this, however, to work around it in core we need to deprecate the old function and return nothing so our hook can do it’s job.

Thats what this PR does. `woocommerce_maybe_show_product_subcategories` calls a new function `woocommerce_output_product_categories` which just does the rendering.

Fixes #18674 

cc @spencerfinnell since you asked about this function the other day.

@claudiulodro @WPprodigy this does need some more testing. It's late here and I may have missed something.

### CODE REVIEW

File: includes/wc-template-functions.php [hunk 1]
```diff
@@ -1830,46 +1830,69 @@ function woocommerce_maybe_show_product_subcategories( $loop_html ) {
...
-				'pad_counts'   => 1,
-			) ) );
-			wp_cache_set( 'product-categories-' . $parent_id, $product_categories, 'product_cat' );
-		}
+		$args = wp_parse_args( $args, $defaults );
 
-		if ( apply_filters( 'woocommerce_product_subcategories_hide_empty', true ) ) {
-			$product_categories = wp_list_filter( $product_categories, array( 'count' => 0 ), 'NOT' );
-		}
+		if ( $args['force_display'] ) {
```
Comment (claudiulodro) [inline includes/wc-template-functions.php]: Why even have this option? :)

If someone is editing their theme to add the special `force_display` arg it's the same amount of work for them as changing the function call to`woocommerce_output_product_categories`.

If you do keep this option, I think the deprecation notice should show when using it. Right now it only shows when not using the force option.

That's all just nitpicking, though.  This works well and I think it's a good solution.

**Review by kloon — APPROVED** [review]
Work for me!

**Comment by DanWrightPhoto** [comment]
Hi, I'm desperate for a work around for this issue as i need to be able to password protect my categories - which i used to be able to do - but obviously can't anymore as my sub cats don't show.
I kind of get what you're on about in this thread.  I'm not a programmer at all.  Is there anywhere i can find a 'walkthrough' so i can edit the code until the new release is out?  Thanks in advance.

**Comment by claudiulodro** [comment]
@DanWrightPhoto this was fixed as part of WooCommerce 3.3.1 back in February. If this is the correct fix for the issue you are experiencing everything should be automatically fixed if you upgrade to WooCommerce version 3.3.1 or newer.

**Pull request merged**